From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Tue, 17 Sep 2024 11:04:19 +0800
Subject: [PATCH 2/3] arm64: errata: Add feature to denote Apple Fusion
 Architecture

The Apple A10 family consists of physical performance and efficiency
cores, and only one of them can be active at a given time depending on
the current p-state. This results in logical cores that could switch
between P-mode and E-mode. This results in the logical core being variable
depending on the current p-state.

For example, only the physical P-core can execute 32-bit EL0, and the
physical cores have different MPIDR values.

Add a feature to denote this situation, and the workarounds for this will
be added later.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 arch/arm64/Kconfig             | 25 +++++++++++++++++++++++++
 arch/arm64/kernel/cpu_errata.c | 30 ++++++++++++++++++++++++++++++
 arch/arm64/tools/cpucaps       |  1 +
 3 files changed, 56 insertions(+)

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index b5a51b0ef..d7db4c493 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -454,6 +454,31 @@ config AMPERE_ERRATUM_AC04_CPU_23
 
 	  If unsure, say Y.
 
+config ARM64_ERRATUM_APPLE_FUSION
+	bool "Apple Hurricane-Zephyr: Variations between physical P-core and E-core presented as a single logical core"
+	depends on ARCH_APPLE
+	default y
+	help
+	  This option disables 32-bit EL0 on Apple A10(X) to workaround
+	  the physical "Zephyr" E-cores not supporting 32-bit EL0.
+
+	  In Apple A10(X), there are physical "Hurricane" performance and
+	  "Zephyr" efficiency cores. They are presented as single cores that
+	  can switch between E-mode and P-mode depending on the current p-state.
+	  Only one of the core types may be active at a given time.
+
+	  However, only the "Hurricane" performance cores supports 32-bit EL0,
+	  so this results in a logical core that can only execute 32-bit EL0
+	  in some states. Apple A10(X) only supports 16KB page sizes so
+	  disabling 32-bit EL0 altogether is acceptable.
+
+	  The P-cores and E-cores also have different MPIDR values, so far
+	  nothing needs to be done, as we do not use code paths that read MPIDR
+	  after boot on A10(X).  However, this is cursed enough such that this
+	  workaround will also taint the kernel for it.
+
+	  If unsure, say Y.
+
 config ARM64_WORKAROUND_CLEAN_CACHE
 	bool
 
diff --git a/arch/arm64/kernel/cpu_errata.c b/arch/arm64/kernel/cpu_errata.c
index 5db8f0619..7bae20b40 100644
--- a/arch/arm64/kernel/cpu_errata.c
+++ b/arch/arm64/kernel/cpu_errata.c
@@ -200,6 +200,20 @@ cpu_enable_cache_maint_trap(const struct arm64_cpu_capabilities *__unused)
 	sysreg_clear_set(sctlr_el1, SCTLR_EL1_UCI, 0);
 }
 
+static void __maybe_unused
+apple_fusion_taint(const struct arm64_cpu_capabilities *__unused)
+{
+	/*
+	 * A variable MPIDR_EL1 would have been a receipe for disaster, but
+	 * since Apple A10, A10X, T2 does not use any of KVM, GIC, ACPI, PSCI
+	 * or cpuidle, and that all CPUs switch between P-mode and E-mode
+	 * together, we do not need to do anything about this (yet). However,
+	 * this is still cursed enough to taint the kernel.
+	 */
+	pr_info_once("Apple Fusion Architecture detected, tainting kernel.\n");
+	add_taint(TAINT_CPU_OUT_OF_SPEC, LOCKDEP_STILL_OK);
+}
+
 #define CAP_MIDR_RANGE(model, v_min, r_min, v_max, r_max)	\
 	.matches = is_affected_midr_range,			\
 	.midr_range = MIDR_RANGE(model, v_min, r_min, v_max, r_max)
@@ -645,6 +659,14 @@ static const struct midr_range erratum_ac04_cpu_23_list[] = {
 };
 #endif
 
+#ifdef CONFIG_ARM64_ERRATUM_APPLE_FUSION
+static const struct midr_range apple_fusion_list[] = {
+	MIDR_ALL_VERSIONS(MIDR_APPLE_A10_T2_HURRICANE_ZEPHYR),
+	MIDR_ALL_VERSIONS(MIDR_APPLE_A10X_HURRICANE_ZEPHYR),
+	{}
+};
+#endif
+
 #ifdef CONFIG_ARM64_WORKAROUND_DISABLE_CNP
 static const struct midr_range cnp_erratum_cpus[] = {
 #ifdef CONFIG_NVIDIA_CARMEL_CNP_ERRATUM
@@ -1008,6 +1030,14 @@ const struct arm64_cpu_capabilities arm64_errata[] = {
 		.capability = ARM64_WORKAROUND_AMPERE_AC04_CPU_23,
 		ERRATA_MIDR_RANGE_LIST(erratum_ac04_cpu_23_list),
 	},
+#endif
+#ifdef CONFIG_ARM64_ERRATUM_APPLE_FUSION
+	{
+		.desc = "Apple Fusion Architecture",
+		.capability = ARM64_WORKAROUND_APPLE_FUSION,
+		ERRATA_MIDR_RANGE_LIST(apple_fusion_list),
+		.cpu_enable = apple_fusion_taint,
+	},
 #endif
 	{
 		.desc = "Broken CNTVOFF_EL2",
diff --git a/arch/arm64/tools/cpucaps b/arch/arm64/tools/cpucaps
index 2775ba335..53cca0616 100644
--- a/arch/arm64/tools/cpucaps
+++ b/arch/arm64/tools/cpucaps
@@ -112,6 +112,7 @@ WORKAROUND_4193714
 WORKAROUND_4311569
 WORKAROUND_AMPERE_AC03_CPU_38
 WORKAROUND_AMPERE_AC04_CPU_23
+WORKAROUND_APPLE_FUSION
 WORKAROUND_TRBE_OVERWRITE_FILL_MODE
 WORKAROUND_TSB_FLUSH_FAILURE
 WORKAROUND_TRBE_WRITE_OUT_OF_RANGE
